File: internal/protocol/v1/signing.go
package v1

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"reflect"
	"unicode/utf8"

	"github.com/gowebpki/jcs"
)

const (
	ControlSigningDomain  = "yuanshu-control-v1\x00"
	OperationDigestDomain = "yuanshu-operation-v1\x00"
)

// ControlSigningInput returns the domain-separated RFC 8785 representation of
// a Protocol v1 control message with the signature member removed. It does not
// validate trust, freshness, replay state, or the signature itself.
func ControlSigningInput(message YuanshuMessage) ([]byte, error) {
	if err := validateControlForEncoding(message); err != nil {
		return nil, err
	}
	if err := validateIJSON(message); err != nil {
		return nil, fmt.Errorf("control message is not I-JSON: %w", err)
	}

	encoded, err := json.Marshal(message)
	if err != nil {
		return nil, fmt.Errorf("marshal control message: %w", err)
	}
	var members map[string]json.RawMessage
	if err := json.Unmarshal(encoded, &members); err != nil {
		return nil, fmt.Errorf("split control message: %w", err)
	}
	delete(members, "signature")
	unsigned, err := json.Marshal(members)
	if err != nil {
		return nil, fmt.Errorf("marshal unsigned control message: %w", err)
	}
	canonical, err := jcs.Transform(unsigned)
	if err != nil {
		return nil, fmt.Errorf("canonicalize control message: %w", err)
	}
	return append([]byte(ControlSigningDomain), canonical...), nil
}

// ApprovalOperationDigest binds the stable target identity and complete
// approval payload while excluding replay-specific event metadata. An existing
// operationDigest member is removed before hashing.
func ApprovalOperationDigest(message YuanshuMessage) (string, error) {
	if err := validateApprovalForDigest(message); err != nil {
		return "", err
	}
	if err := validateIJSON(message); err != nil {
		return "", fmt.Errorf("approval message is not I-JSON: %w", err)
	}

	payloadJSON, err := json.Marshal(message.Payload)
	if err != nil {
		return "", fmt.Errorf("marshal approval payload: %w", err)
	}
	var payload map[string]json.RawMessage
	if err := json.Unmarshal(payloadJSON, &payload); err != nil {
		return "", fmt.Errorf("split approval payload: %w", err)
	}
	delete(payload, "operationDigest")
	payloadJSON, err = json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("marshal digest payload: %w", err)
	}

	binding := approvalDigestBinding{
		ProtocolVersion: message.ProtocolVersion,
		Type:            message.Type,
		OwnerID:         message.OwnerID,
		NodeID:          message.NodeID,
		WorkspaceID:     message.WorkspaceID,
		ThreadID:        message.ThreadID,
		TurnID:          message.TurnID,
		ItemID:          message.ItemID,
		Payload:         payloadJSON,
	}
	encoded, err := json.Marshal(binding)
	if err != nil {
		return "", fmt.Errorf("marshal approval binding: %w", err)
	}
	canonical, err := jcs.Transform(encoded)
	if err != nil {
		return "", fmt.Errorf("canonicalize approval binding: %w", err)
	}
	hashInput := append([]byte(OperationDigestDomain), canonical...)
	digest := sha256.Sum256(hashInput)
	return base64.RawURLEncoding.EncodeToString(digest[:]), nil
}

type approvalDigestBinding struct {
	ProtocolVersion string          `json:"protocolVersion"`
	Type            string          `json:"type"`
	OwnerID         string          `json:"ownerId"`
	NodeID          string          `json:"nodeId"`
	WorkspaceID     *string         `json:"workspaceId,omitempty"`
	ThreadID        *string         `json:"threadId,omitempty"`
	TurnID          *string         `json:"turnId,omitempty"`
	ItemID          *string         `json:"itemId,omitempty"`
	Payload         json.RawMessage `json:"payload"`
}

func validateControlForEncoding(message YuanshuMessage) error {
	if message.ProtocolVersion != CurrentVersion {
		return fmt.Errorf("unsupported control protocol version %q", message.ProtocolVersion)
	}
	if !IsKnownControl(message.Type) {
		return fmt.Errorf("unsupported control type %q", message.Type)
	}
	if message.MessageID == "" || message.SentAt == "" || message.OwnerID == "" || message.NodeID == "" || message.StreamID == "" || message.CorrelationID == "" {
		return errors.New("control message is missing a required envelope identifier")
	}
	if message.ExpiresAt == nil || *message.ExpiresAt == "" || message.Nonce == nil || *message.Nonce == "" || message.Signer == nil || message.Signer.ClientID == "" || message.Signer.KeyID == "" || message.Payload == nil {
		return errors.New("control message is missing required signing fields")
	}
	if message.Sequence < 0 || message.Sequence > 9007199254740991 {
		return errors.New("control sequence is outside the JavaScript safe-integer range")
	}
	return nil
}

func validateApprovalForDigest(message YuanshuMessage) error {
	if message.ProtocolVersion != CurrentVersion || message.Type != string(EventApprovalRequested) {
		return errors.New("operation digest requires a Protocol v1 approval.requested event")
	}
	if message.OwnerID == "" || message.NodeID == "" || message.Payload == nil {
		return errors.New("approval event is missing a required binding field")
	}
	approvalID, approvalOK := message.Payload["approvalId"].(string)
	kind, kindOK := message.Payload["kind"].(string)
	if !approvalOK || approvalID == "" || !kindOK || kind == "" {
		return errors.New("approval payload requires non-empty approvalId and kind")
	}
	return nil
}

func validateIJSON(value any) error {
	return validateJSONValue(reflect.ValueOf(value), make(map[visit]bool))
}

type visit struct {
	typeName reflect.Type
	pointer  uintptr
}

func validateJSONValue(value reflect.Value, seen map[visit]bool) error {
	if !value.IsValid() {
		return nil
	}
	if value.Kind() == reflect.Interface || value.Kind() == reflect.Pointer {
		if value.IsNil() {
			return nil
		}
		if value.Kind() == reflect.Pointer {
			key := visit{value.Type(), value.Pointer()}
			if seen[key] {
				return errors.New("cyclic value")
			}
			seen[key] = true
			defer delete(seen, key)
		}
		return validateJSONValue(value.Elem(), seen)
	}

	switch value.Kind() {
	case reflect.String:
		if !utf8.ValidString(value.String()) {
			return errors.New("invalid UTF-8 string")
		}
	case reflect.Float32, reflect.Float64:
		number := value.Float()
		if math.IsNaN(number) || math.IsInf(number, 0) {
			return errors.New("non-finite number")
		}
	case reflect.Map:
		if value.Type().Key().Kind() != reflect.String {
			return errors.New("JSON object key is not a string")
		}
		if value.IsNil() {
			return nil
		}
		key := visit{value.Type(), value.Pointer()}
		if seen[key] {
			return errors.New("cyclic value")
		}
		seen[key] = true
		defer delete(seen, key)
		iterator := value.MapRange()
		for iterator.Next() {
			if !utf8.ValidString(iterator.Key().String()) {
				return errors.New("invalid UTF-8 object key")
			}
			if err := validateJSONValue(iterator.Value(), seen); err != nil {
				return err
			}
		}
	case reflect.Slice:
		if value.IsNil() {
			return nil
		}
		key := visit{value.Type(), value.Pointer()}
		if seen[key] {
			return errors.New("cyclic value")
		}
		seen[key] = true
		defer delete(seen, key)
		for index := 0; index < value.Len(); index++ {
			if err := validateJSONValue(value.Index(index), seen); err != nil {
				return err
			}
		}
	case reflect.Array:
		for index := 0; index < value.Len(); index++ {
			if err := validateJSONValue(value.Index(index), seen); err != nil {
				return err
			}
		}
	case reflect.Struct:
		for index := 0; index < value.NumField(); index++ {
			if value.Type().Field(index).PkgPath != "" {
				continue
			}
			if err := validateJSONValue(value.Field(index), seen); err != nil {
				return err
			}
		}
	case reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return nil
	default:
		return fmt.Errorf("unsupported JSON value kind %s", value.Kind())
	}
	return nil
}

File: web/src/protocol/v1/signing.ts
import canonicalize from "canonicalize";
import { CURRENT_VERSION, KNOWN_CONTROL_TYPES } from "./catalog.generated";
import type { YuanshuMessage } from "./types.generated";

export const CONTROL_SIGNING_DOMAIN = "yuanshu-control-v1\0";
export const OPERATION_DIGEST_DOMAIN = "yuanshu-operation-v1\0";

const encoder = new TextEncoder();
const knownControls = new Set<string>(KNOWN_CONTROL_TYPES);

export function controlSigningInput(message: YuanshuMessage): Uint8Array {
  validateControlForEncoding(message);
  assertIJson(message);
  const { signature: _signature, ...unsigned } = message;
  return domainSeparated(CONTROL_SIGNING_DOMAIN, canonicalJson(unsigned));
}

export async function approvalOperationDigest(message: YuanshuMessage): Promise<string> {
  validateApprovalForDigest(message);
  assertIJson(message);
  const { operationDigest: _digest, ...payload } = message.payload;
  const binding: Record<string, unknown> = {
    protocolVersion: message.protocolVersion,
    type: message.type,
    ownerId: message.ownerId,
    nodeId: message.nodeId,
    payload,
  };
  for (const field of ["workspaceId", "threadId", "turnId", "itemId"] as const) {
    if (message[field] !== undefined) binding[field] = message[field];
  }
  const input = domainSeparated(OPERATION_DIGEST_DOMAIN, canonicalJson(binding));
  const digest = await crypto.subtle.digest("SHA-256", input.slice().buffer as ArrayBuffer);
  return bytesToBase64Url(new Uint8Array(digest));
}

function validateControlForEncoding(message: YuanshuMessage): void {
  if (message.protocolVersion !== CURRENT_VERSION) throw new Error(`unsupported control protocol version ${message.protocolVersion}`);
  if (!knownControls.has(message.type)) throw new Error(`unsupported control type ${message.type}`);
  for (const field of ["messageId", "sentAt", "expiresAt", "ownerId", "nodeId", "streamId", "correlationId", "nonce"] as const) {
    if (typeof message[field] !== "string" || message[field].length === 0) throw new Error(`control message is missing ${field}`);
  }
  if (!message.signer || !message.signer.clientId || !message.signer.keyId) throw new Error("control message is missing signer");
  if (!Number.isSafeInteger(message.sequence) || message.sequence < 0) throw new Error("control sequence is outside the JavaScript safe-integer range");
  if (!isPlainObject(message.payload)) throw new Error("control payload must be an object");
}

function validateApprovalForDigest(message: YuanshuMessage): void {
  if (message.protocolVersion !== CURRENT_VERSION || message.type !== "approval.requested") {
    throw new Error("operation digest requires a Protocol v1 approval.requested event");
  }
  if (!message.ownerId || !message.nodeId || !isPlainObject(message.payload)) throw new Error("approval event is missing a required binding field");
  if (typeof message.payload.approvalId !== "string" || message.payload.approvalId.length === 0 || typeof message.payload.kind !== "string" || message.payload.kind.length === 0) {
    throw new Error("approval payload requires non-empty approvalId and kind");
  }
}

function canonicalJson(value: unknown): string {
  const result = canonicalize(value);
  if (result === undefined) throw new Error("value cannot be represented as canonical JSON");
  return result;
}

function domainSeparated(domain: string, canonical: string): Uint8Array {
  const prefix = encoder.encode(domain);
  const body = encoder.encode(canonical);
  const result = new Uint8Array(prefix.length + body.length);
  result.set(prefix);
  result.set(body, prefix.length);
  return result;
}

function assertIJson(value: unknown, seen = new Set<object>()): void {
  if (value === null || typeof value === "boolean") return;
  if (typeof value === "string") {
    if (!isWellFormedUnicode(value)) throw new Error("I-JSON string contains an unpaired surrogate");
    return;
  }
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new Error("I-JSON number must be finite");
    return;
  }
  if (typeof value !== "object") throw new Error(`unsupported I-JSON value type ${typeof value}`);
  if (seen.has(value)) throw new Error("cyclic value");
  seen.add(value);
  try {
    if (Array.isArray(value)) {
      for (const item of value) assertIJson(item, seen);
      return;
    }
    if (!isPlainObject(value)) throw new Error("I-JSON objects must use a plain object prototype");
    for (const [key, item] of Object.entries(value)) {
      if (!isWellFormedUnicode(key)) throw new Error("I-JSON object key contains an unpaired surrogate");
      assertIJson(item, seen);
    }
  } finally {
    seen.delete(value);
  }
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return false;
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

function isWellFormedUnicode(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const unit = value.charCodeAt(index);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (!(next >= 0xdc00 && next <= 0xdfff)) return false;
      index += 1;
    } else if (unit >= 0xdc00 && unit <= 0xdfff) {
      return false;
    }
  }
  return true;
}

function bytesToBase64Url(bytes: Uint8Array): string {
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replaceAll("+", "-").replaceAll("/", "_").replace(/=+$/, "");
}
